feat: Show how to embed libpython in a Rust Wasm module (#105)
vmware-labs/webassembly-language-runtimes: python/examples/embedding/wasi-command-rs/README.md

# About

A simple WASI Command (exports `_start`) Wasm module, written in Rust.

Embeds CPython via libpython and demonstrates interaction with simple Python code via [pyo3](https://pyo3.rs/v0.19.0/).

# How to run

Make sure you have `cargo` with the `wasm32-wasi` target. For running we use `wasmtime`, but the module will work with any WASI-compliant runtime.

Just run `./run_me.sh` in the current folder. You will see something like this

```
wlr/python/examples/embedding/wasi-command-rs $$ ./run_me.sh
   Compiling pyo3-build-config v0.18.3
   ...
    Finished dev [unoptimized + debuginfo] target(s) in 26.43s
Hello from Python(libpython3.11.a) in Wasm(Rust). args= ('a1', 'a2', 3, 4)
```

# About the code

This example is really simple and adopted from `pyo3`'s documentation on [calling Python from Rust](https://pyo3.rs/v0.18.3/python_from_rust).

> It's main purpose is to show how to configure the build and dependencies.

Reviewer: ```suggestion
Its main purpose is to show how to configure the build and dependencies.
```

# Build and dependencies

For pyo3 to work the final binary needs to link to `libpython3.11.a`. The WLR project provides a pre-build `libpython` static library, which depends on `wasi-sdk`. To setup the build properly you will need to provide several static libs and configure the linker to use them properly.

> The build uses pre-built `wasm32-wasi` static libs, based on [wasi-sdk](https://github.com/WebAssembly/wasi-sdk). There is custom code in [build.rs](./build.rs), which downloads them and configures the linker to properly use them.

Reviewer 2: Shall we make more obvious we're providing it?

```suggestion
Fortunately, the custom code in [build.rs](./build.rs) downloads the required libraries and configures the linker to use them for you. The build uses pre-built `wasm32-wasi` static libs, based on [wasi-sdk](https://github.com/WebAssembly/wasi-sdk).
```
